File: backend/src/main/java/com/rgcet/admission/service/CutoffCalculator.java
package com.rgcet.admission.service;

import com.rgcet.admission.entity.HSCAcademicMark;
import com.rgcet.admission.entity.HSCVocationalMark;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.Objects;

public final class CutoffCalculator {

    public static final BigDecimal CUTOFF_MAX_SCORE = BigDecimal.valueOf(300);

    private static final List<String> SCIENCE_SUBJECTS =
            List.of("Chemistry", "Computer Science", "Biology", "Bio Technology", "Biotechnology");

    private CutoffCalculator() {
    }

    /**
     * Engineering Cut-Off = Maths + Physics + best of (Chemistry / Computer Science / Biology / Bio Technology).
     * Each subject is normalised to its percentage (marksObtained / maximumMarks * 100) before summing,
     * so subjects marked out of any maximum contribute a score out of 100. The maximum possible cutoff is 300.
     */
    public static BigDecimal engineeringCutOff(List<HSCAcademicMark> academicMarks) {
        if (academicMarks == null || academicMarks.isEmpty()) {
            return null;
        }
        BigDecimal maths = subjectPercentage(academicMarks, "Maths", "Mathematics");
        BigDecimal physics = subjectPercentage(academicMarks, "Physics");
        if (maths == null || physics == null) {
            return null;
        }
        BigDecimal bestScience = SCIENCE_SUBJECTS.stream()
                .map(s -> subjectPercentage(academicMarks, s))
                .filter(Objects::nonNull)
                .max(BigDecimal::compareTo)
                .orElse(null);
        if (bestScience == null) {
            return null;
        }
        return maths.add(physics).add(bestScience);
    }

    /** Merit percentage used for fee slab lookup: engineeringCutOff / 300 * 100. */
    public static BigDecimal meritPercent(BigDecimal engineeringCutOff) {
        if (engineeringCutOff == null || CUTOFF_MAX_SCORE.signum() == 0) {
            return null;
        }
        return engineeringCutOff.multiply(BigDecimal.valueOf(100))
                .divide(CUTOFF_MAX_SCORE, 2, RoundingMode.HALF_UP);
    }

    public static BigDecimal subjectPercentage(BigDecimal obtained, BigDecimal maximum) {
        if (obtained == null || maximum == null || maximum.signum() == 0) {
            return null;
        }
        return obtained.multiply(BigDecimal.valueOf(100))
                .divide(maximum, 2, RoundingMode.HALF_UP);
    }

    public static BigDecimal overallPercentage(List<HSCAcademicMark> academic,
                                               List<HSCVocationalMark> vocational) {
        BigDecimal obtained = totalMarksObtained(academic, vocational);
        BigDecimal maximum = totalMaxMarks(academic, vocational);
        return subjectPercentage(obtained, maximum);
    }

    public static BigDecimal totalMaxMarks(List<HSCAcademicMark> academic,
                                           List<HSCVocationalMark> vocational) {
        BigDecimal total = BigDecimal.ZERO;
        boolean any = false;
        if (academic != null) {
            for (HSCAcademicMark m : academic) {
                if (m.getMaximumMarks() != null) {
                    total = total.add(m.getMaximumMarks());
                    any = true;
                }
            }
        }
        if (vocational != null) {
            for (HSCVocationalMark m : vocational) {
                if (m.getMaximumMarks() != null) {
                    total = total.add(m.getMaximumMarks());
                    any = true;
                }
            }
        }
        return any ? total : null;
    }

    public static BigDecimal totalMarksObtained(List<HSCAcademicMark> academic,
                                                List<HSCVocationalMark> vocational) {
        BigDecimal total = BigDecimal.ZERO;
        boolean any = false;
        if (academic != null) {
            for (HSCAcademicMark m : academic) {
                if (m.getMarksObtained() != null) {
                    total = total.add(m.getMarksObtained());
                    any = true;
                }
            }
        }
        if (vocational != null) {
            for (HSCVocationalMark m : vocational) {
                if (m.getMarksObtained() != null) {
                    total = total.add(m.getMarksObtained());
                    any = true;
                }
            }
        }
        return any ? total : null;
    }

    private static BigDecimal subjectPercentage(List<HSCAcademicMark> academicMarks, String... names) {
        for (HSCAcademicMark mark : academicMarks) {
            if (mark.getSubjectName() == null) {
                continue;
            }
            for (String name : names) {
                if (mark.getSubjectName().equalsIgnoreCase(name)) {
                    return subjectPercentage(mark.getMarksObtained(), mark.getMaximumMarks());
                }
            }
        }
        return null;
    }
}

File: src/utils/cutoffCalculator.ts
import { HSCSubjectMark } from '../types';

/**
 * New Cut-Off Formula (out of 300):
 * Cut-Off = Maths (100) + Physics (100) + 3rd Subject (100)
 * Where 3rd Subject = Biology / Computer Science / Bio Technology / Chemistry
 */
export const calculateHSCCutOff = (marks: HSCSubjectMark[], stream: 'Academic' | 'Vocational'): number => {
  if (!marks || marks.length === 0) return 0;

  const getPercentage = (mark: HSCSubjectMark): number => {
    if (!mark || !mark.maxMarks || mark.maxMarks <= 0) return 0;
    return Number(((mark.marksObtained / mark.maxMarks) * 100).toFixed(2));
  };

  if (stream === 'Academic') {
    // Formula: Maths (100) + Physics (100) + 3rd Subject (100) = 300
    const maths = marks.find((m) => m.subject.toLowerCase().includes('maths'));
    const physics = marks.find((m) => m.subject.toLowerCase().includes('physics'));
    // 3rd subject: Biology, Computer Science, Bio Technology, or Chemistry
    const thirdSubject = marks.find((m) =>
      m.subject.toLowerCase().includes('biology') ||
      m.subject.toLowerCase().includes('computer') ||
      m.subject.toLowerCase().includes('bio technology') ||
      m.subject.toLowerCase().includes('biotechnology') ||
      m.subject.toLowerCase().includes('chemistry')
    );

    const mathsPct = maths ? getPercentage(maths) : 0;
    const physicsPct = physics ? getPercentage(physics) : 0;
    const thirdPct = thirdSubject ? getPercentage(thirdSubject) : 0;

    const cutoff = mathsPct + physicsPct + thirdPct;
    return Number(cutoff.toFixed(2));
  } else {
    // Vocational Formula: Vocational Subject (100) + Related I (100) + Related II (100) = 300
    const vocational = marks.find((m) => m.subject.toLowerCase().includes('vocational'));
    const related1 = marks.find((m) => m.subject.toLowerCase().includes('related subject i'));
    const related2 = marks.find((m) => m.subject.toLowerCase().includes('related subject ii'));

    const vocPct = vocational ? getPercentage(vocational) : 0;
    const r1Pct = related1 ? getPercentage(related1) : 0;
    const r2Pct = related2 ? getPercentage(related2) : 0;

    const cutoff = vocPct + r1Pct + r2Pct;
    return Number(cutoff.toFixed(2));
  }
};
